lib/openstudio-standards/utilities/assertion.rb

```diff
@@ -0,0 +1,135 @@
+# PRM generation custom Error
+class PRMError < StandardError
+end
+
+# PRM assertion method
+# Raise assertion if the test (bool) is failed
+#
+# Before raise the exception, the method will generate the prm.log for debugging
+# @param bool [Boolean, Object] an object for truthy evaluation
+# @param log_dir [String] log file directory
+# @param err_msg [String] message raise the exception
+def prm_raise(bool, log_dir, err_msg)
+  unless bool
+    # OpenStudio.logFree(OpenStudio::Debug, 'prm.log', log_msg)
+    log_messages_to_file_prm("#{log_dir}/prm.log", true)
+    raise PRMError, "#{err_msg} - Check debug log at #{log_dir}/prm.log for more information"
+  end
+end
+
+# PRM reading function reads user data from a hash map.
+# Handles key existence, value is nil and value string is empty
+# @param user_data [Hash] a hash contains a user data
+# @param key [String] key string
+# @param default [Object] values assigned if the data is not available.
+# @return data [String] a string
+def prm_read_user_data(user_data, key, default = nil)
+  return user_data.key?(key) && !user_data[key].nil? && !user_data[key].to_s.empty? ? user_data[key] : default
+end
+
+# This is a PRM handler function handles the .get from an optional object
+# The handler will try to access the OpenStudio Object data key
+# And do it recursively until all the keys have been checked and final object get or raise exception
+# for non_initialized objects.
+# @param component [OpenStudio::Model::Component] an OpenStudio object
+# @param log_dir [string] directory to save the log
+# @param data_key [string] The data key to retrieve the data from the OpenStudio object
+# @param remaining_keys [str] Any additional keys in the path
+# @return the OpenStudio Object or exception raise
+def prm_get_optional_handler(component, log_dir, data_key, *remaining_keys)
```

Comment:
Why is this function necessary?

Reply:
We have many situations to get optional object from an OpenStudio object.
```Ruby
spacetype = space.spaceType.get.standardsSpaceType.get
```
In above code, if a space has no spaceType or standardsSpaceType, than ruby will raise exception with message `nil has no get function`. This happens quite often with models generated from BEM software. The error message is not informative.

The `prm_get_optional_handler` is designed to handle such scenario to give more informative error messages to user.

Reply:
The better practice is to call .is_initialized each time, and then have comments at that point in code if the optional value is not initialized.

Search .is_initialized in the codebase and you'll see many examples of if - else statements to catch this sort of thing.

Reply:
Right, but there are multiple chain options in the codebase so this method comes in as convenience to handle those chain optional get.
I can add .is_initialized in the `prm_get_optional_handler` method to reflect the best practice?

Reply:
I see 20 uses of this method in the code base, and 1 instance of chaining, which I elsewhere commented wasn't necessary.
I think this method doesn't accomplish what you want - giving more clear error messages to the user.

Reply:
In the PRM routine, there are growing number of chaining for processing user data including some of them pointing to `name`.
Examples include:
[Get building user data name](https://github.com/NREL/openstudio-standards/blob/AppendixG_Dev/lib/openstudio-standards/standards/ashrae_90_1_prm/ashrae_90_1_prm.Model.rb#L1748) 
[get space types:](https://github.com/NREL/openstudio-standards/blob/AppendixG_Dev/lib/openstudio-standards/standards/ashrae_90_1_prm/ashrae_90_1_prm.Model.rb#L2403)

Most of them are specific to PRM user data processing so I would like, if possible, to keep it for now.

I can update the error message to make the issue clearer to the user?